Gerrit issue945

jedkcanderson opened the pull request:
https://github.com/zaggino/brackets-git/issues/945

@nmorenor, myself, and the rest of our team have been working tag-team on this issue as we transition from Java development into Brackets hacking.  It's kinda fun having a handful of us on a hangout as we hack & slash JS and CSS code!

We believe we have solved your performance concern by using a preference.  

We thought that brackets would have "native" support for a checked menu item but we couldn't find one in the dropdown.css so we borrowed the CSS from the Bracket's filetype menu.  Is this something we should bring up to the brackets team?  

We have a couple of questions that the code didn't answer for us:
Are preference constants typically inlined?  Or is it common to define them in a file somewhere?  We saw them inlined, but we didn't know if this is standard or the result of organic code growth.
We copied the error handling when we can't write the config file from another spot in the code.  Is this considered "good enough"?  Can we do better?

zaggino commented:
Hi @jedkcanderson , first of all - nice work :+1: 

Preferences which are inlined in `Preferences.js` are real preferences and you can override them per project in your `.brackets.json` like this one https://github.com/zaggino/brackets-git/blob/master/.brackets.json

If you don't put your preference into `Preferences.js` it'll be just a state, something you don't want user to see or use in his config files.

zaggino commented on src/Panel.js:
```diff
@@ -674,8 +674,27 @@ define(function (require, exports) {
         // Clone button
         $gitPanel.find(".git-clone").prop("disabled", false);
 
+        var p3 = Git.getConfig("gerrit.pushref").then(function (enabled) {
```
Does this need to happen on every refresh? Only the panel contents should be required to refresh, not the buttons and the menu. There is probably a better place to move this to.

zaggino replied:
I can help find the correct place later, I'm just doing quick comments now.

zaggino replied:
This should be probably hooked on `GIT_ENABLED` event. How often do you need to get this executed or what was the motivation for putting it here?

See this section lower (line 1073):
```
EventEmitter.on(Events.GIT_ENABLED, function () {
        // Add info from Git to panel
        Git.getConfig("user.name").then(function (currentUserName) {
            EventEmitter.emit(Events.GIT_USERNAME_CHANGED, currentUserName);
        });
        Git.getConfig("user.email").then(function (currentEmail) {
            EventEmitter.emit(Events.GIT_EMAIL_CHANGED, currentEmail);
        });
    });
```

zaggino commented:
Looks good, just one concern which I'll look into a little later.

zaggino commented on src/Panel.js:
```diff
@@ -674,8 +674,27 @@ define(function (require, exports) {
+        var p3 = Git.getConfig("gerrit.pushref").then(function (enabled) {
+            var gerritPreference = Preferences.get("gerrit.pushref");
+            if (enabled === "true" && (gerritPreference === undefined || !gerritPreference)) {
+                // Handles the case where we switched to a repo that is using gerrit
```
`GIT_ENABLED` is called on every branch and project change.

zaggino commented on src/Panel.js:
```diff
@@ -882,6 +901,37 @@ define(function (require, exports) {
+            // so we don't slow down people who aren't using gerrit.
+            Preferences.persist("gerrit.pushref", true);
+        }
+        return Git.getConfig("gerrit.pushref").then(function (enabled) {
```
call the output var strEnabled to avoid confusion

zaggino commented on src/Panel.js:
```diff
@@ -882,6 +901,37 @@ define(function (require, exports) {
+    EventEmitter.on(Events.GERRIT_TOGGLE_PUSH_REF, function (event, callback) {
+        // update menu item state
+        var gerritPushref = Preferences.get("gerrit.pushref");
+        if (gerritPushref === undefined || !gerritPushref) {
```
comparation to undefined is not needed when you have !something, !undefined is also true

zaggino commented on src/Panel.js:
```diff
@@ -674,8 +674,27 @@ define(function (require, exports) {
+        return Promise.all([p1, p2, p3]);
+    }
+
+    function setGerritCheckState(enabled) {
```
refactor so this method is always called with boolean, not string

jedkcanderson commented:
From your comment about preferences, I think we should add the new preference key this to the Preferences.js so that users can commit the setting.  If one person on a team needs the setting, it's likely that they all will.

We put the call to read the preference in the refresh() method because we wanted the setting to be read when the user switched projects. And if we do add this to the Preferences.js, then a revert could trigger a .brackets.json file change which would then cause the setting to need to be read again.  That being said, we're open to suggestions for moving the call.

zaggino commented on src/Panel.js:
```diff
@@ -882,6 +901,37 @@ define(function (require, exports) {
+            Preferences.persist("gerrit.pushref", true);
+        }
+        return Git.getConfig("gerrit.pushref").then(function (enabled) {
+            if (enabled === "true") {
```
this should be `var enabled = strEnabled === "true"`

zaggino commented:
1) yes, define an entry in preferences.js
2) use GIT_ENABLED event, it will execute on project/branch change

zaggino commented:
sorry for spamming all over the place, im reviewing this from my mobile :)

jedkcanderson commented:
I believe my latest push addresses your concerns regarding which event to listen for, using booleans as parameters, and making a user-settable preference.

zaggino commented:
Thanks @jedkcanderson, looks great. Will merge this and probably do a release to the registry later today.

jedkcanderson commented:
Sounds good.  Thanks!

zaggino commented:
merged :+1: really good work guys
